File: leetcode/Clone_Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 *
 * 
 * Solution: this clone is based on the edge instead of the node. 
 *           This way will make sure the edge will be copied to the
 *           cloned graph as well.
 *
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if(node == nullptr) return nullptr;
        
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> map;
        stack<UndirectedGraphNode *> st;
        UndirectedGraphNode *t = nullptr;
        UndirectedGraphNode *nGraph = new UndirectedGraphNode(node->label);
        map[node] = nGraph;
        
        st.push(node);
        while(!st.empty()){
            t = st.top();
            st.pop();
            
            for(int i = 0; i < t->neighbors.size(); i++){
                UndirectedGraphNode *nb = t->neighbors[i];
                if(map.find(nb) == map.end()){
                    UndirectedGraphNode *x = new UndirectedGraphNode(nb->label);
                    map[nb] = x;
                    st.push(nb);
                }
                
                map[t]->neighbors.push_back(map[nb]);
            }
        }
        
        return nGraph;
    }
};


/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 * 
 * 
 * BFS based solution.
 * It is almost identical to the DFS solution except the order of edge copying is different from DFS.
 * DFS will make go deeper and deeper, while BFS will copy edges layer by layer.
 *
 * Every time a node is pushed into queue, its clone is created. Also their mapping relationship is
 * kept in map<UndirectedGraphNode *, UndirectedGraphNode *> to clone edge later.
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if(node == nullptr) return nullptr;
        
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> map;
        queue<UndirectedGraphNode *> queue;
        UndirectedGraphNode *t = nullptr;
        UndirectedGraphNode *nGraph = new UndirectedGraphNode(node->label);
        map[node] = nGraph;
        
        queue.push(node);
        while(!queue.empty()){
            t = queue.front();
            queue.pop();
            
            for(int i = 0; i < t->neighbors.size(); i++){
                UndirectedGraphNode *nb = t->neighbors[i];
                if(map.find(nb) == map.end()){
                    UndirectedGraphNode *x = new UndirectedGraphNode(nb->label);
                    map[nb] = x;
                    queue.push(nb);
                }
               
                // for each edge, clone a copy on the new graph 
                map[t]->neighbors.push_back(map[nb]);
            }
        }
        
        return nGraph;
    }
};

File: leetcode/java/CloneGraph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */

/* Graph is assumed to be connected, and can be reached through DFS/BFS */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if(node == null) return null;
        
        UndirectedGraphNode newNode = new UndirectedGraphNode(node.label);
        Map<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<>();
        map.put(node, newNode);
        
        Queue<UndirectedGraphNode> queue = new ArrayDeque<>();
        queue.offer(node);
        while(!queue.isEmpty()) {
            UndirectedGraphNode x = queue.poll();
            for(UndirectedGraphNode nb : x.neighbors) {
                if(!map.containsKey(nb)) {
                    UndirectedGraphNode y = new UndirectedGraphNode(nb.label);
                    map.put(nb, y);
                    queue.offer(nb);
                }
                map.get(x).neighbors.add(map.get(nb));
            }
        }
        return newNode;
    }
}


// DFS
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        
        Deque<Node> st = new ArrayDeque<>();
        Map<Node, Node> m = new HashMap<>();
        st.push(node);
        m.put(node, new Node(node.val));
        
        while(!st.isEmpty()) {
            Node x = st.pop();
            for(Node t : x.neighbors){
                if(!m.containsKey(t)) {
                    Node p = new Node(t.val);
                    m.put(t, p);
                    st.push(t);
                }
                m.get(x).neighbors.add(m.get(t));
            }
        }
        return m.get(node);
    }
}
